pandas/tools/merge.py

```diff
@@ -273,9 +273,17 @@ def _get_join_info(self):
                                                  sort=self.sort, how=self.how)
 
             if self.right_index:
-                join_index = self.left.index.take(left_indexer)
+                if len(self.left) > 0:
+                    join_index = self.left.index.take(left_indexer)
+                else:
+                    join_index = self.right.index.take(right_indexer)
+                    left_indexer = np.array([np.nan] * len(join_index))
             elif self.left_index:
-                join_index = self.right.index.take(right_indexer)
+                if len(self.right) > 0:
+                    join_index = self.right.index.take(right_indexer)
+                else:
+                    join_index = self.left.index.take(left_indexer)
+                    right_indexer = np.array([np.nan] * len(join_index))
```

Comment:
OK, change indexer to ``int (-1)``, and added ``dtype`` checks (done by ``assert_frame_equal`` by default).